src/rapids_singlecell/preprocessing/_sparse_pca/_dask_sparse_pca.py

```diff
@@ -0,0 +1,198 @@
+        When False, the local gram matrix result
+        will be overwritten
+    return_mean: boolean (default = False)
+        If True, the Maximum Likelihood Estimate used to
+        calculate the mean of X and X will be returned,
+        of the form (1 / n) * mean(X) and (1 / n) * mean(X)
+
+    Returns
+    -------
+
+    result : cov(X, X) when return_gram and return_mean are False
+            cov(X, X), gram(X, X) when return_gram is True,
+            return_mean is False
+            cov(X, X), mean(X), mean(X) when return_gram is False,
+            return_mean is True
+            cov(X, X), gram(X, X), mean(X), mean(X)
+            when return_gram is True and return_mean is True
+    """
+    import dask
+
+    from ._kernels._pca_sparse_kernel import (
+        _copy_kernel,
+        _cov_kernel,
+        _gramm_kernel_csr,
+    )
+
+    compute_mean_cov = _gramm_kernel_csr(x.dtype)
+    compute_mean_cov.compile()
+    n_cols = x.shape[1]
+
+    def __gram_block(x_part):
+        gram_matrix = cp.zeros((n_cols, n_cols), dtype=x.dtype)
+
+        block = (128,)
+        grid = (x_part.shape[0],)
+        compute_mean_cov(
+            grid,
+            block,
+            (
+                x_part.indptr,
+                x_part.indices,
+                x_part.data,
+                x_part.shape[0],
+                n_cols,
+                gram_matrix,
+            ),
+        )
+        return gram_matrix[None, ...]  # need new axis for summing
+
+    n_blocks = len(x.to_delayed().ravel())
+    gram_matrix = x.map_blocks(
+        __gram_block,
+        new_axis=(1,),
+        chunks=((1,) * n_blocks, (x.shape[1],), (x.shape[1],)),
+        meta=cp.array([]),
+        dtype=x.dtype,
+    ).sum(axis=0)
+    mean_x, _ = _get_mean_var(x)
+    gram_matrix, mean_x = dask.compute(gram_matrix, mean_x)
+    mean_x = mean_x.astype(x.dtype)
+    copy_gram = _copy_kernel(x.dtype)
+    block = (32, 32)
+    grid = (math.ceil(n_cols / block[0]), math.ceil(n_cols / block[1]))
+    copy_gram(
+        grid,
+        block,
+        (gram_matrix, n_cols),
+    )
+
+    gram_matrix *= 1 / x.shape[0]
+
+    if return_gram:
+        cov_result = cp.zeros(
+            (gram_matrix.shape[0], gram_matrix.shape[0]),
+            dtype=gram_matrix.dtype,
+        )
+    else:
+        cov_result = gram_matrix
+
+    compute_cov = _cov_kernel(gram_matrix.dtype)
+
+    block_size = (32, 32)
+    grid_size = (math.ceil(gram_matrix.shape[0] / 8),) * 2
+    compute_cov(
+        grid_size,
+        block_size,
+        (cov_result, gram_matrix, mean_x, mean_x, gram_matrix.shape[0]),
+    )
+
+    if not return_gram and not return_mean:
+        return cov_result
+    elif return_gram and not return_mean:
+        return cov_result, gram_matrix
+    elif not return_gram and return_mean:
+        return cov_result, mean_x, mean_x
+    elif return_gram and return_mean:
+        return cov_result, gram_matrix, mean_x, mean_x
```

Comment:
Just a thing I noticed while reading this code: Why return the same thing twice?

Reply:
answer: because Severin wants to upstream this, and it’s the upstream API